### PR TITLE
refactor(sync): use task scheduler

### PR DESCRIPTION
**Pull request by peusebiu**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/controller.go [hunk 1]
```diff
@@ -48,6 +49,7 @@ type Controller struct {
 	Server          *http.Server
 	Metrics         monitoring.MetricServer
 	CveInfo         ext.CveInfo
+	SyncOnDemand    SyncOnDemand
```
Comment (rchincha) [inline pkg/api/controller.go:52]: This is going backwards, extensions is leaking into controller.

Comment (peusebiu) [reply pkg/api/controller.go:52]: like the above CveInfo and Metrics, right? :D

Comment (rchincha) [reply pkg/api/controller.go:52]: everything :(

File: pkg/extensions/sync/local.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+						Err(err).Msg("couldn't upload manifest because of missing annotations")
+
+					return nil
+				}
+
+				return err
+			}
+		}
+
+		_, _, err = imageStore.PutImageManifest(repo, reference, mediaType, manifestBlob)
```
Comment (andaaron) [inline pkg/extensions/sync/local.go:165]: No changes needed here for signatures?

File: pkg/extensions/sync/utils.go [hunk 3]
```diff
@@ -1,251 +1,40 @@
...
 	"github.com/containers/image/v5/docker"
-	"github.com/containers/image/v5/docker/reference"
 	"github.com/containers/image/v5/manifest"
-	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/signature"
 	"github.com/containers/image/v5/types"
-	guuid "github.com/gofrs/uuid"
-	godigest "github.com/opencontainers/go-digest"
+	"github.com/docker/distribution/reference"
+	"github.com/opencontainers/go-digest"
```
Comment (andaaron) [inline pkg/extensions/sync/utils.go:20]: Let's keep `godigest` as an alias. It's less confusing when accessing the apis within the package.

File: pkg/extensions/sync/sync_test.go [hunk 4]
```diff
@@ -787,28 +846,35 @@ func TestOnDemand(t *testing.T) {
...
 
-			// repodb fails for syncCosignSignature"
+			dcm.StopServer()
 
+			// repodb fails for syncOCIRefs
 			dctlr.RepoDB = mocks.RepoDBMock{
-				SetRepoReferenceFn: func(repo, reference string, manifestDigest godigest.Digest, mediaType string) error {
-					if strings.HasPrefix(reference, "sha256") || strings.HasSuffix(reference, ".sig") {
+				SetRepoReferenceFn: func(repo, Reference string, manifestDigest godigest.Digest, mediaType string) error {
+					if mediaType == ispec.MediaTypeArtifactManifest {
```
Comment (andaaron) [inline pkg/extensions/sync/sync_test.go:861]: Please double check this logic is still relevant today.

File: pkg/extensions/sync/sync_test.go [hunk 5]
```diff
@@ -2141,56 +2308,55 @@ func TestNotSemver(t *testing.T) {
 	})
 }
 
-func TestErrorOnCatalog(t *testing.T) {
-	Convey("Verify error on catalog", t, func() {
-		updateDuration, _ := time.ParseDuration("1h")
+// func TestErrorOnCatalog(t *testing.T) {
```
Comment (andaaron) [inline pkg/extensions/sync/sync_test.go:2311]: This needs an update.

File: pkg/extensions/sync/on_demand.go [hunk 6]
```diff
@@ -81,281 +67,103 @@ func OneImage(ctx context.Context, cfg syncconf.Config, repoDB repodb.RepoDB,
...
-	defer demandedImgs.delete(demandedImage)
-	defer close(imageChannel)
+	syncResult := make(chan error)
+	onDemand.requestStore.Store(req, syncResult)
+
+	defer onDemand.requestStore.Delete(req)
+	defer close(syncResult)
 
-	go syncOneImage(ctx, imageChannel, cfg, repoDB, storeController, repo, reference, artifactType, log)
+	go onDemand.syncImage(repo, reference, syncResult)
```
Comment (andaaron) [inline pkg/extensions/sync/on_demand.go:76]: Why do we need this goroutine?

**Review by andaaron — COMMENTED** [review]
Overall I like the new structure. It will be a big improvement.

**Review by andaaron — COMMENTED** [review]
Overall I like the new structure. It will be a big improvement.

File: pkg/common/common_test.go [hunk 7]
```diff
@@ -115,9 +115,9 @@ func TestCommon(t *testing.T) {
 		var resultPtr interface{}
 		httpClient, err := common.CreateHTTPClient(true, "localhost", tempDir)
 		So(err, ShouldBeNil)
-		_, _, err = common.MakeHTTPGetRequest(httpClient, "", "",
+		_, _, _, err = common.MakeHTTPGetRequest(httpClient, "", "",
 			resultPtr, baseURL+"/v2/", ispec.MediaTypeImageManifest, log.NewLogger("", ""))
-		So(err, ShouldNotBeNil)
+		So(err, ShouldBeNil)
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/common/common_test.go:120]: now I use MakeHTTPGetRequest to ping registries (GET on /v2/ route) so I changed the function to not Unmarshal if there is no body on the response, so now getting on /v2/ does not return an Unmarshal error anymore.

[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/common/common_test.go:120]: Ok, thanks.

File: pkg/api/routes.go [hunk 8]
```diff
@@ -1708,36 +1696,47 @@ func getImageManifest(ctx context.Context, routeHandler *RouteHandler, imgStore
 		routeHandler.c.Log.Info().Str("repository", name).Str("reference", reference).
 			Msg("trying to get updated image by syncing on demand")
 
-		errSync := ext.SyncOneImage(ctx, routeHandler.c.Config, routeHandler.c.RepoDB, routeHandler.c.StoreController,
-			name, reference, "", routeHandler.c.Log)
-		if errSync != nil {
-			routeHandler.c.Log.Err(errSync).Str("repository", name).Str("reference", reference).
-				Msg("error encounter while syncing image")
+		if isCosignTag(reference) {
+			if errSync := routeHandler.c.SyncOnDemand.SyncReference(name, reference, syncConstants.Cosign); errSync != nil {
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/routes.go:1700]: what do you mean? I use SyncOnDemand.SyncReference() for notation and cosign signatures, and oci artifacts and SyncOnDemand.SyncImage() for image use case. 

Do you mean to make the call for everything on the same function?

File: pkg/extensions/sync/on_demand.go (hunk 6 above)
Comment (peusebiu) [reply pkg/extensions/sync/on_demand.go:76]: We only start this go routine on the first call to the outer function, the rest of simultaneous calls are waiting for the first one to exit

File: pkg/extensions/sync/local.go (hunk 2 above)
Comment (peusebiu) [reply pkg/extensions/sync/local.go:165]: no, at the moment, cosign tags are skipped from this logic, and notation signatures don't have tags so they are also skipped from this logic.

File: pkg/extensions/sync/utils.go [hunk 9]
```diff
@@ -1,251 +1,40 @@
...
-
-	if repoDestination == "/" {
-		return "/"
-	}
+const (
+	SyncBlobUploadDir = ".sync"
+)
 
-	return strings.Trim(repoDestination, "/")
+type Credentials struct {
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/extensions/sync/utils.go:35]: yes, it's not used, just a leftover, will remove, thanks

**Review by andaaron — APPROVED** [review]
LGTM

**Review by andaaron — APPROVED** [review]

**Review by andaaron — APPROVED** [review]

**Review by andaaron — DISMISSED** [review]
LGTM

File: pkg/api/routes.go [hunk 10]
```diff
@@ -488,31 +486,26 @@ type ImageIndex struct {
...
 	artifactTypes []string,
 ) (ispec.Index, error) {
-	references, err := imgStore.GetReferrers(name, digest, artifactTypes)
-	if err != nil || len(references.Manifests) == 0 {
-		if routeHandler.c.Config.Extensions != nil &&
-			routeHandler.c.Config.Extensions.Sync != nil &&
-			*routeHandler.c.Config.Extensions.Sync.Enable {
+	refs, err := imgStore.GetReferrers(name, digest, artifactTypes)
+	if err != nil || len(refs.Manifests) == 0 {
+		if isSyncOnDemandEnable(*routeHandler.c) {
```
Comment (rchincha) [inline pkg/api/routes.go:495]: "Enabled"?

File: pkg/common/common.go [hunk 11]
```diff
@@ -208,20 +210,21 @@ func MakeHTTPGetRequest(httpClient *http.Client, username string, password strin
 		log.Error().Str("status code", fmt.Sprint(resp.StatusCode)).
 			Err(err).Str("blobURL", blobURL).Msg("couldn't get blob")
 
-		return nil, resp.StatusCode, errors.New(string(body)) //nolint:goerr113
+		return nil, "", resp.StatusCode, errors.New(string(body)) //nolint:goerr113
 	}
 
 	// read blob
+	if len(body) > 0 {
```
Comment (rchincha) [inline pkg/common/common.go:217]: if resp.ContentLength > 0 also?

File: pkg/extensions/extension_sync.go [hunk 12]
```diff
@@ -4,33 +4,48 @@
...
-		if err := sync.Run(ctx, *config.Extensions.Sync, repoDB, storeController, log); err != nil {
-			log.Error().Err(err).Msg("Error encountered while setting up syncing")
+		onDemandService := sync.NewOnDemandService(log)
+
+		for _, registryConfig := range config.Extensions.Sync.Registries {
+			isPeriodical := len(registryConfig.Content) != 0 && registryConfig.PollInterval != 0
+			isOnDemand := registryConfig.OnDemand
+
+			if isPeriodical || isOnDemand {
+				service, err := sync.NewService(registryConfig, config.Extensions.Sync.CredentialsFile,
```
Comment (rchincha) [inline pkg/extensions/extension_sync.go:26]: "sync.New()"

Comment (rchincha) [reply pkg/extensions/extension_sync.go:26]: per registryConfig

File: pkg/extensions/extension_sync.go [hunk 13]
```diff
@@ -4,33 +4,48 @@
...
-func EnableSyncExtension(ctx context.Context, config *config.Config,
-	repoDB repodb.RepoDB, storeController storage.StoreController, log log.Logger,
-) {
+func EnableSyncExtension(config *config.Config, repoDB repodb.RepoDB,
+	storeController storage.StoreController, sch *scheduler.Scheduler, log log.Logger,
+) (*sync.BaseOnDemand, error) {
 	if config.Extensions.Sync != nil && *config.Extensions.Sync.Enable {
-		if err := sync.Run(ctx, *config.Extensions.Sync, repoDB, storeController, log); err != nil {
-			log.Error().Err(err).Msg("Error encountered while setting up syncing")
+		onDemandService := sync.NewOnDemandService(log)
```
Comment (rchincha) [inline pkg/extensions/extension_sync.go:19]: "NewOndemand()"

drop "Service" unless this is a long-running service

File: pkg/extensions/extension_sync.go [hunk 14]
```diff
@@ -4,33 +4,48 @@
...
+			if isPeriodical || isOnDemand {
+				service, err := sync.NewService(registryConfig, config.Extensions.Sync.CredentialsFile,
+					storeController, repoDB, log)
+				if err != nil {
+					return nil, err
+				}
+
+				if isPeriodical {
+					// add to task scheduler periodic sync
+					gen := sync.NewGenerator(service, log)
```
Comment (rchincha) [inline pkg/extensions/extension_sync.go:34]: NewTaskGenerator()

File: pkg/extensions/extension_sync.go [hunk 15]
```diff
@@ -4,33 +4,48 @@
...
+					sch.SubmitGenerator(gen, registryConfig.PollInterval, scheduler.MediumPriority)
```
Comment (rchincha) [inline pkg/extensions/extension_sync.go:35]: SubmitTaskGenerator()?

File: pkg/extensions/extension_sync.go [hunk 16]
```diff
@@ -4,33 +4,48 @@
...
+				}
+
+				if isOnDemand {
+					// onDemand services used in routes.go
+					onDemandService.AddService(service)
```
Comment (rchincha) [inline pkg/extensions/extension_sync.go:40]: onDemand.Add()

File: pkg/extensions/sync/content.go [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+	"zotregistry.io/zot/pkg/common"
+	syncconf "zotregistry.io/zot/pkg/extensions/config/sync"
+	"zotregistry.io/zot/pkg/log"
+)
+
+type ContentManager struct {
+	contents []syncconf.Content
+	log      log.Logger
+}
+
```
Comment (rchincha) [inline pkg/extensions/sync/content.go:22]: Add comments on what it does

File: pkg/extensions/sync/sync.go [hunk 18]
```diff
@@ -1,412 +1,110 @@
...
-		SourceCtx:             upstreamCtx,
-		ReportWriter:          io.Discard,
-		ForceManifestMIMEType: ispec.MediaTypeImageManifest, // force only oci manifest MIME type
-		ImageListSelection:    copy.CopyAllImages,
-	}
-
-	return options
+type Local interface {
+	CanSkipImage(repo, tag string, imageDigest digest.Digest) (bool, error)
+	CommitImage(imageReference types.ImageReference, repo, tag string) error
```
Comment (rchincha) [inline pkg/extensions/sync/sync.go:47]: Add comments in all interfaces

File: pkg/extensions/sync/sync.go [hunk 19]
```diff
@@ -1,412 +1,110 @@
...
 	}
+}
 
-	registryURL, err := url.Parse(upstreamURL)
-	if err != nil {
-		log.Error().Str("errorType", common.TypeOf(err)).
-			Err(err).Str("url", upstreamURL).Msg("couldn't parse url")
-
-		return err
+func (gen *TaskGenerator) GenerateTask() (scheduler.Task, error) {
```
Comment (rchincha) [inline pkg/extensions/sync/sync.go:67]: s/GenerateTask()/Next()

or NextTask()?

Comment (rchincha) [reply pkg/extensions/sync/sync.go:67]: Add comments on how the generator is working

**Comment by rchincha** [comment]
When syncing to an existing local repo, we may clobber "latest", do sane things and log

**Comment by rchincha** [comment]
For a given/existing repo, sync and clients may both be writers to the same repo, so locking?

**Comment by rchincha** [comment]
Recursive download of references (I expect two levels)

File: test/blackbox/sync_docker.bats [hunk 20]
```diff
@@ -23,17 +23,43 @@ function setup_file() {
...
     "extensions": {
         "sync": {
             "registries": [
                 {
                     "urls": [
-                        "https://docker.io/library",
-                        "https://registry.k8s.io",
-                        "https://aws.amazon.com/ecr",
-                        "https://gcr.io",
+                        "https://docker.io/library"
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply test/blackbox/sync_docker.bats:34]: solved with content filtering

File: pkg/api/routes.go [hunk 21]
```diff
@@ -1745,36 +1733,45 @@ func getImageManifest(ctx context.Context, routeHandler *RouteHandler, imgStore
...
+					Msg("error encounter while syncing image")
+			}
 		}
 	}
 
 	return imgStore.GetImageManifest(name, reference)
 }
 
+func isCosignTag(tag string) bool {
+	if strings.HasPrefix(tag, "sha256-") && strings.HasSuffix(tag, remote.SignatureTagSuffix) {
```
Comment (andaaron) [inline pkg/api/routes.go:1754]: The SBOM suffix should be tested for as well.

**Review by andaaron — APPROVED** [review]

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**